File: stations/_archive/hook-observatory-py/handlers/relay_signal.py
"""
tmux relay signal — Stop handler.

When Claude Code finishes a response, checks if a tmux-relay is waiting
for this pane's completion signal. If so, sends `tmux wait-for -S`.
Also updates Redis cache with pane status and pre-caches result.
Designed to be fast (<10ms) — no-op when no relay is pending.
"""

from __future__ import annotations

import json
import os
import time

from .base import ALLOW, HookResult, run_background

# Lazy Redis connection — only initialized when needed
_redis_client = None


def _get_redis():
    global _redis_client
    if _redis_client is None:
        try:
            import redis

            _redis_client = redis.from_url(
                os.getenv("WORKSHOP_REDIS_URL", "redis://localhost:6379/0"),
                decode_responses=True,
            )
        except Exception:
            return None
    return _redis_client


def _update_redis_cache(pane_safe: str) -> None:
    """Update Redis cache: mark pane idle + pre-cache result."""
    r = _get_redis()
    if not r:
        return

    try:
        raw = r.hget("relay:panes", pane_safe)
        signal_file = ""
        if raw:
            data = json.loads(raw)
            signal_file = data.get("signal_file", "")
            data["status"] = "idle"
            data["updated_at"] = time.time()
            r.hset("relay:panes", pane_safe, json.dumps(data))

        # Pre-cache result
        if signal_file:
            basename = os.path.basename(signal_file)
            r.set(
                f"relay:result:{basename}",
                json.dumps(
                    {
                        "status": "success",
                        "completed_at": time.time(),
                        "pane": f"%{pane_safe}",
                    }
                ),
                ex=3600,
            )

        r.set("relay:cache_ts", str(time.time()), ex=60)
    except Exception:
        pass  # Cache is advisory — never block the hook


def _debug_log(msg: str) -> None:
    """Append debug line to relay signal log (advisory, never fails)."""
    try:
        with open("/tmp/relay-signal-debug.log", "a") as f:
            f.write(f"{time.strftime('%H:%M:%S')} {msg}\n")
    except Exception:
        pass


def handle(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    pane_id = os.environ.get("TMUX_PANE", "")
    if not pane_id:
        return ALLOW

    pane_safe = pane_id.replace("%", "")
    pending_file = f"/tmp/relay-pending-{pane_safe}.channel"

    if not os.path.isfile(pending_file):
        _debug_log(f"no-op pane={pane_safe} (no pending file)")
        return ALLOW

    try:
        with open(pending_file) as f:
            channel = f.read().strip()
    except OSError:
        _debug_log(f"error pane={pane_safe} (cannot read pending file)")
        return ALLOW

    if channel:
        try:
            os.remove(pending_file)
        except OSError:
            pass
        run_background(["tmux", "wait-for", "-S", channel])
        _debug_log(f"signaled pane={pane_safe} channel={channel}")

        # Update Redis cache (advisory, non-blocking)
        _update_redis_cache(pane_safe)

    return ALLOW

File: internal/handlers/relay_signal.go
package handlers

// relay_signal.go — Go port of handlers/relay_signal.py
//
// Stop handler: checks if a tmux-relay is waiting for this pane, signals via
// `tmux wait-for -S`, and updates Redis pane cache.

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	rclient "github.com/joneshong/hook-dispatcher/internal/clients"
	"github.com/joneshong/hook-dispatcher/internal/core"
)

func init() {
	core.Register("Stop", core.Entry{
		Matcher:    "",
		Handler:    relaySignalHandle,
		Critical:   false,
		ModuleName: "relay_signal",
	})
}

func relaySignalHandle(_, _ string, _ map[string]any, _ string) core.HookResult {
	paneID := os.Getenv("TMUX_PANE")
	if paneID == "" {
		return core.Allow()
	}
	pane := strings.TrimPrefix(paneID, "%")
	pendingFile := fmt.Sprintf("/tmp/relay-pending-%s.channel", pane)

	if _, err := os.Stat(pendingFile); os.IsNotExist(err) {
		relayDebugLog(fmt.Sprintf("no-op pane=%s (no pending file)", pane))
		return core.Allow()
	}

	data, err := os.ReadFile(pendingFile)
	if err != nil {
		relayDebugLog(fmt.Sprintf("error pane=%s (cannot read pending file)", pane))
		return core.Allow()
	}

	channel := strings.TrimSpace(string(data))
	if channel == "" {
		return core.Allow()
	}

	// Remove pending file before signalling to avoid double-fire.
	_ = os.Remove(pendingFile)
	_ = core.RunBackground([]string{"tmux", "wait-for", "-S", channel}, "")
	relayDebugLog(fmt.Sprintf("signaled pane=%s channel=%s", pane, channel))

	// Advisory Redis cache update — fail-open.
	relayUpdateRedisCache(pane)

	return core.Allow()
}

func relayUpdateRedisCache(pane string) {
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()

	raw, err := rclient.HGet(ctx, "relay:panes", pane)
	if err != nil || raw == "" {
		return
	}

	var info map[string]any
	if err := json.Unmarshal([]byte(raw), &info); err != nil {
		return
	}

	signalFile, _ := info["signal_file"].(string)
	info["status"] = "idle"
	info["updated_at"] = time.Now().Unix()

	updated, err := json.Marshal(info)
	if err == nil {
		_ = rclient.HSet(ctx, "relay:panes", pane, string(updated))
	}

	if signalFile != "" {
		basename := filepath.Base(signalFile)
		result := map[string]any{
			"status":       "success",
			"completed_at": time.Now().Unix(),
			"pane":         "%" + pane,
		}
		resultJSON, _ := json.Marshal(result)
		_ = rclient.Set(ctx, "relay:result:"+basename, string(resultJSON), time.Hour)
	}

	_ = rclient.Set(ctx, "relay:cache_ts", fmt.Sprintf("%f", float64(time.Now().UnixNano())/1e9), time.Minute)
}

func relayDebugLog(msg string) {
	f, err := os.OpenFile("/tmp/relay-signal-debug.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return
	}
	defer f.Close()
	ts := time.Now().Format("15:04:05")
	_, _ = fmt.Fprintf(f, "%s %s\n", ts, msg)
}
